crates/db/src/schema.rs

```diff
@@ -79,6 +79,8 @@ pub const COLUMN_BLOCK_SUBMIT_TX_HASH: Col = 7;
 pub const COLUMN_BLOCK_DEPOSIT_INFO_VEC: Col = 16;
 /// block number (in big endian) -> FinalizedCustodianCapacity.
 pub const COLUMN_BLOCK_POST_FINALIZED_CUSTODIAN_CAPACITY: Col = 36;
+/// block hash -> finalizing block number range
+pub const COLUMN_BLOCK_FINALIZING_RANGE: Col = 37;
```

Comment:
I kinda feel that this may not be necessary. L2 block timestamp is ascending. We can use binary or maybe [exponential search](https://en.wikipedia.org/wiki/Exponential_search) to find the last finalized block.

Or maybe we can use binary or exponential search when this is not stored so we don't need a migration (or re-syncing from genesis).

Reply:
It is good to find the last finalized via search. It's like a space-time trade, I think.


I came up with another search method: start a cursor at `tip_number - rollup_config.finality_blocks`, and then go up or go down according to the result of `is_finalized(tip, cursor)`.
This method builds based on the assumption that `rollup_config.finality_as_duration` is roughly equal to `rollup_config.finality_blocks` block intervals.

Reply:
The new commit d269603b608f734ac7a21e8054a46d1aa589d53b implement `calc_finalizing_range` via searching. PTAL.
The main changes were made to these two files:
- https://github.com/keroro520/godwoken/blob/new-finality-rule-based-on-timestamp/crates/utils/src/calc_finalizing_range.rs
- https://github.com/keroro520/godwoken/blob/new-finality-rule-based-on-timestamp/crates/tests/src/tests/calc_finalizing_range.rs